src/v/archival/ntp_archiver_service.cc

```diff
@@ -0,0 +1,290 @@
+    s3::client client(_client_conf);
+    batch_result result{};
+    auto fut = ss::parallel_for_each(
+      upload_set,
+      [this, &lm, &result](manifest::segment_map::value_type target) {
+          return upload_segment(std::move(target), lm).then([&result](bool ok) {
+              if (ok) {
+                  result.succeded++;
+              } else {
+                  result.failed++;
+              }
+          });
+      });
+    co_await std::move(fut);
+    if (num_candidates) {
+        vlog(archival_log.trace, "Completed S3 upload");
+        co_await client.shutdown();
+        co_await upload_manifest();
+        _last_upload_time = ss::lowres_clock::now();
+    }
+    co_return result;
+}
+
+ss::future<bool> ntp_archiver::delete_segment(
+  manifest::segment_map::value_type target, storage::log_manager& lm) {
+    const auto& [sname, meta] = target;
+    auto segment = get_segment(sname, lm);
+    if (!segment) {
+        s3::client client(_client_conf);
+        auto s3path = _remote.get_remote_segment_path(sname);
+        vlog(archival_log.trace, "Delete segment \"{}\" from S3", s3path());
+        try {
+            co_await client.delete_object(_bucket, s3::object_key(s3path()));
+            _remote.delete_permanently(sname);
+        } catch (...) {
+            vlog(
+              archival_log.error,
+              "Failed to delete {} from S3. Reason: {}",
+              sname,
+              std::current_exception());
+            co_return false;
+        }
+    }
+    co_return true;
+}
+
+ss::future<ntp_archiver::batch_result> ntp_archiver::delete_next_candidate(
+  ss::semaphore& req_limit, storage::log_manager& lm) {
+    vlog(archival_log.trace, "Delete next candidate called");
+    gate_guard guard{_gate};
+    // calculate candidate set
+    auto local = _policy->make_local_manifest(lm);
+    if (!local) {
+        vlog(archival_log.error, "Failed to generate local manifest");
+        co_return batch_result{};
+    }
+    auto candidates = _remote.difference(*local);
+    auto num_candidates = std::min(
+      static_cast<ssize_t>(candidates.size()), req_limit.available_units());
+    if (num_candidates == 0) {
+        co_return batch_result{};
+    }
+    std::vector<manifest::segment_map::value_type> delete_set;
+    std::copy_n(
+      candidates.begin(), num_candidates, std::back_inserter(delete_set));
+    vlog(archival_log.trace, "Delete {} elements", delete_set.size());
+    // upload segments in parallel
+    batch_result result{};
+    s3::client client(_client_conf);
+    auto fut = ss::parallel_for_each(
+      delete_set,
+      [this, &lm, &result](manifest::segment_map::value_type target) {
+          return delete_segment(std::move(target), lm).then([&result](bool ok) {
+              if (ok) {
+                  result.succeded++;
+              } else {
+                  result.failed++;
+              }
+          });
+      });
+    co_await std::move(fut);
```

Comment:
what is the reason for this? related to the linking issue? do you have a reproducer for that linking issue? for example, this seems to work ok which seems similar.
```
std::vector<int> asdf;
co_await ss::parallel_for_each(asdf, [](int){ return ss::now().then([](){});});
```